Provider StandardRuleSetProvider not a subtype

Hi! 
I was trying to migrate to most recent version of `kotlinter` and for most cases it worked as expected, However, there are 2 cases when it fails:
1. When it's been run on a buildserver which performs completely clean build.
2. If kotlinter's tasks are run with `--rerun-tasks` argument. (e.g. `./gradlew lintKotlin --rerun-tasks`)

## Callstack:
```
> Task :analytics:lintKotlinMain FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':analytics:lintKotlinMain'.
> There were multiple failures while executing work items
   > A failure occurred while executing org.jmailen.gradle.kotlinter.tasks.lint.LintWorkerRunnable
      > com.pinterest.ktlint.core.RuleSetProvider: Provider com.pinterest.ktlint.ruleset.standard.StandardRuleSetProvider not a subtype
   > A failure occurred while executing org.jmailen.gradle.kotlinter.tasks.lint.LintWorkerRunnable
      > com.pinterest.ktlint.core.RuleSetProvider: Provider com.pinterest.ktlint.ruleset.standard.StandardRuleSetProvider not a subtype
   > A failure occurred while executing org.jmailen.gradle.kotlinter.tasks.lint.LintWorkerRunnable
      > com.pinterest.ktlint.core.RuleSetProvider: Provider com.pinterest.ktlint.ruleset.standard.StandardRuleSetProvider not a subtype
   > A failure occurred while executing org.jmailen.gradle.kotlinter.tasks.lint.LintWorkerRunnable
      > com.pinterest.ktlint.core.RuleSetProvider: Provider com.pinterest.ktlint.ruleset.standard.StandardRuleSetProvider not a subtype
   > A failure occurred while executing org.jmailen.gradle.kotlinter.tasks.lint.LintWorkerRunnable
      > com.pinterest.ktlint.core.RuleSetProvider: Provider com.pinterest.ktlint.ruleset.standard.StandardRuleSetProvider not a subtype

* Try:
Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':analytics:lintKotlinMain'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$3.accept(ExecuteActionsTaskExecuter.java:151)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$3.accept(ExecuteActionsTaskExecuter.java:148)
        at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:191)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:141)
        at org.gradle.api.internal.tasks.execution.ResolveBeforeExecutionStateTaskExecuter.execute(ResolveBeforeExecutionStateTaskExecuter.java:75)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:108)
        at org.gradle.api.internal.tasks.execution.ResolveBeforeExecutionOutputsTaskExecuter.execute(ResolveBeforeExecutionOutputsTaskExecuter.java:67)
        at org.gradle.api.internal.tasks.execution.ResolveAfterPreviousExecutionStateTaskExecuter.execute(ResolveAfterPreviousExecutionStateTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:94)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:95)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:73)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:49)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:416)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:406)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:102)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:49)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:43)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:355)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:343)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:336)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:322)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker$1.execute(DefaultPlanExecutor.java:134)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker$1.execute(DefaultPlanExecutor.java:129)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:202)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:193)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:129)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
Caused by: org.gradle.workers.WorkerExecutionException: There were multiple failures while executing work items
        at org.gradle.workers.internal.DefaultWorkerExecutor.workerExecutionException(DefaultWorkerExecutor.java:160)
        at org.gradle.workers.internal.DefaultWorkerExecutor.await(DefaultWorkerExecutor.java:152)
        at org.jmailen.gradle.kotlinter.tasks.LintTask.run(LintTask.kt:84)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:103)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.doExecute(StandardTaskAction.java:48)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:41)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:28)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:702)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:669)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$5.run(ExecuteActionsTaskExecuter.java:404)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:402)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:394)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:92)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:393)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:376)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.access$200(ExecuteActionsTaskExecuter.java:80)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$TaskExecution.execute(ExecuteActionsTaskExecuter.java:213)
        at org.gradle.internal.execution.steps.ExecuteStep.lambda$execute$1(ExecuteStep.java:33)
        at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:33)
        at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:26)
        at org.gradle.internal.execution.steps.CleanupOutputsStep.execute(CleanupOutputsStep.java:58)
        at org.gradle.internal.execution.steps.CleanupOutputsStep.execute(CleanupOutputsStep.java:35)
        at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:48)
        at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:33)
        at org.gradle.internal.execution.steps.CancelExecutionStep.execute(CancelExecutionStep.java:39)
        at org.gradle.internal.execution.steps.TimeoutStep.executeWithoutTimeout(TimeoutStep.java:73)
        at org.gradle.internal.execution.steps.TimeoutStep.execute(TimeoutStep.java:54)
        at org.gradle.internal.execution.steps.CatchExceptionStep.execute(CatchExceptionStep.java:35)
        at org.gradle.internal.execution.steps.CreateOutputsStep.execute(CreateOutputsStep.java:51)
        at org.gradle.internal.execution.steps.SnapshotOutputsStep.execute(SnapshotOutputsStep.java:45)
        at org.gradle.internal.execution.steps.SnapshotOutputsStep.execute(SnapshotOutputsStep.java:31)
        at org.gradle.internal.execution.steps.CacheStep.executeWithoutCache(CacheStep.java:201)
        at org.gradle.internal.execution.steps.CacheStep.execute(CacheStep.java:70)
        at org.gradle.internal.execution.steps.CacheStep.execute(CacheStep.java:45)
        at org.gradle.internal.execution.steps.BroadcastChangingOutputsStep.execute(BroadcastChangingOutputsStep.java:49)
        at org.gradle.internal.execution.steps.StoreSnapshotsStep.execute(StoreSnapshotsStep.java:43)
        at org.gradle.internal.execution.steps.StoreSnapshotsStep.execute(StoreSnapshotsStep.java:32)
        at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:38)
        at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:24)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.executeBecause(SkipUpToDateStep.java:96)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.lambda$execute$0(SkipUpToDateStep.java:89)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:54)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:38)
        at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:77)
        at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:37)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:36)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:26)
        at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:90)
        at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:48)
        at org.gradle.internal.execution.impl.DefaultWorkExecutor.execute(DefaultWorkExecutor.java:33)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:120)
        ... 35 more
Caused by: org.gradle.workers.internal.DefaultWorkerExecutor$WorkExecutionException: A failure occurred while executing org.jmailen.gradle.kotlinter.tasks.lint.LintWorkerRunnable
        at org.gradle.workers.internal.DefaultWorkerExecutor$WorkerExecution.waitForCompletion(DefaultWorkerExecutor.java:285)
        at org.gradle.internal.work.DefaultAsyncWorkTracker.waitForItemsAndGatherFailures(DefaultAsyncWorkTracker.java:115)
        at org.gradle.internal.work.DefaultAsyncWorkTracker.waitForCompletion(DefaultAsyncWorkTracker.java:87)
        at org.gradle.workers.internal.DefaultWorkerExecutor.await(DefaultWorkerExecutor.java:150)
        ... 88 more
Caused by: java.util.ServiceConfigurationError: com.pinterest.ktlint.core.RuleSetProvider: Provider com.pinterest.ktlint.ruleset.standard.StandardRuleSetProvider not a subtype
        at org.jmailen.gradle.kotlinter.support.RuleSetsKt.resolveRuleSets(ruleSets.kt:25)
        at org.jmailen.gradle.kotlinter.support.RuleSetsKt.resolveRuleSets$default(ruleSets.kt:11)
        at org.jmailen.gradle.kotlinter.tasks.lint.LintWorkerRunnable.run(LintWorkerRunnable.kt:47)
        at org.gradle.workers.internal.DefaultWorkerServer.execute(DefaultWorkerServer.java:39)
        at org.gradle.workers.internal.NoIsolationWorkerFactory$1$1.execute(NoIsolationWorkerFactory.java:61)
        at org.gradle.workers.internal.AbstractWorker$1.call(AbstractWorker.java:44)
        at org.gradle.workers.internal.AbstractWorker$1.call(AbstractWorker.java:41)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:416)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:406)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:102)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
        at org.gradle.workers.internal.AbstractWorker.executeWrappedInBuildOperation(AbstractWorker.java:41)
        at org.gradle.workers.internal.NoIsolationWorkerFactory$1.execute(NoIsolationWorkerFactory.java:55)
        at org.gradle.workers.internal.DefaultWorkerExecutor$1.call(DefaultWorkerExecutor.java:105)
        at org.gradle.workers.internal.DefaultWorkerExecutor$1.call(DefaultWorkerExecutor.java:99)
        at org.gradle.internal.work.DefaultConditionalExecutionQueue$ExecutionRunner.runExecution(DefaultConditionalExecutionQueue.java:215)
        at org.gradle.internal.work.DefaultConditionalExecutionQueue$ExecutionRunner.runBatch(DefaultConditionalExecutionQueue.java:164)
        at org.gradle.internal.work.DefaultConditionalExecutionQueue$ExecutionRunner.run(DefaultConditionalExecutionQueue.java:131)
        ... 3 more
```
 
## Current setup:
Gradle 5.4.1, Kotlin 1.3.31 (tested on 1.3.21 too), Kotlinter 1.25.0.
I'm applying the plugin within `buildsrc` by setting:
```
    pluginManager.apply("org.jmailen.kotlinter")
    extensions.configure(KotlinterExtension::class.java) {
        it.indentSize = 4
        it.continuationIndentSize = 4
        it.reporters = arrayOf("plain")
    }
```
in custom buildSrc plugin.

While I'll be still trying to reproduce the issue on a new project, ~~I'm reporting this at sight cause I assume that can be fixed on the library side as after downgrading library version down to *1.23.1* everything works like a charm. The issue has been introduced in release `1.24.0`~~

EDIT: Apparently the issue has been introduced earlier, in `1.23.0`. Tested on sample project attached below.

--- a/README.md
+++ b/README.md
@@ -15,7 +15,7 @@ Available on the Gradle Plugins Portal: https://plugins.gradle.org/plugin/org.jm
 
 ```kotlin
 plugins {
-    id("org.jmailen.kotlinter") version "1.25.0"
+    id("org.jmailen.kotlinter") version "1.25.1"
 }
 ```
 
@@ -26,7 +26,7 @@ plugins {
 
 ```groovy
 plugins {
-    id "org.jmailen.kotlinter" version "1.25.0"
+    id "org.jmailen.kotlinter" version "1.25.1"
 }
 ```
 
@@ -46,7 +46,7 @@ buildscript {
         }
     }
     dependencies {
-        classpath("org.jmailen.gradle:kotlinter-gradle:1.25.0")
+        classpath("org.jmailen.gradle:kotlinter-gradle:1.25.1")
     }
 }
 ```
@@ -71,7 +71,7 @@ buildscript {
         }
     }
     dependencies {
-        classpath "org.jmailen.gradle:kotlinter-gradle:1.25.0"
+        classpath "org.jmailen.gradle:kotlinter-gradle:1.25.1"
     }
 }
 ```

--- a/build.gradle
+++ b/build.gradle
@@ -3,7 +3,7 @@ plugins {
     id 'com.gradle.plugin-publish' version '0.10.0'
     id 'java-gradle-plugin'
     id 'maven-publish'
-    id 'org.jmailen.kotlinter' version '1.24.1'
+    id 'org.jmailen.kotlinter' version '1.25.0'
     id 'idea'
 }
 
@@ -30,7 +30,7 @@ tasks.withType(PluginUnderTestMetadata).configureEach {
     pluginClasspath.from(configurations.compileOnly)
 }
 
-version = '1.25.0'
+version = '1.25.1'
 group = 'org.jmailen.gradle'
 def pluginId = 'org.jmailen.kotlinter'
 

--- a/src/main/kotlin/org/jmailen/gradle/kotlinter/support/ruleSets.kt
+++ b/src/main/kotlin/org/jmailen/gradle/kotlinter/support/ruleSets.kt
@@ -7,8 +7,8 @@ import java.util.ServiceLoader
 import kotlin.comparisons.compareBy
 
 fun resolveRuleSets(
-    includeExperimentalRules: Boolean = false,
-    providers: Iterable<RuleSetProvider> = ServiceLoader.load(RuleSetProvider::class.java)
+    providers: Iterable<RuleSetProvider>,
+    includeExperimentalRules: Boolean = false
 ): List<RuleSet> {
     return providers
         .filter { includeExperimentalRules || it !is ExperimentalRuleSetProvider }
@@ -20,3 +20,5 @@ fun resolveRuleSets(
             }
         })
 }
+
+fun defaultProviders() = ServiceLoader.load(RuleSetProvider::class.java)

--- a/src/main/kotlin/org/jmailen/gradle/kotlinter/tasks/FormatTask.kt
+++ b/src/main/kotlin/org/jmailen/gradle/kotlinter/tasks/FormatTask.kt
@@ -7,7 +7,7 @@ import org.gradle.api.tasks.TaskAction
 import org.gradle.workers.WorkerExecutor
 import org.jmailen.gradle.kotlinter.KotlinterExtension
 import org.jmailen.gradle.kotlinter.support.ExecutionContextRepository
-import org.jmailen.gradle.kotlinter.support.resolveRuleSets
+import org.jmailen.gradle.kotlinter.support.defaultProviders
 import org.jmailen.gradle.kotlinter.tasks.format.FormatExecutionContext
 import org.jmailen.gradle.kotlinter.tasks.format.FormatWorkerConfigurationAction
 import org.jmailen.gradle.kotlinter.tasks.format.FormatWorkerParameters
@@ -40,9 +40,8 @@ open class FormatTask @Inject constructor(
 
     @TaskAction
     fun run() {
-        val ruleSets = resolveRuleSets(experimentalRules)
         val executionContextRepository = ExecutionContextRepository.formatInstance
-        val executionContext = FormatExecutionContext(ruleSets, logger)
+        val executionContext = FormatExecutionContext(defaultProviders(), logger)
         val executionContextRepositoryId = executionContextRepository.register(executionContext)
 
         source

--- a/src/main/kotlin/org/jmailen/gradle/kotlinter/tasks/FormatTask.kt
+++ b/src/main/kotlin/org/jmailen/gradle/kotlinter/tasks/FormatTask.kt
@@ -7,6 +7,7 @@ import org.gradle.api.tasks.TaskAction
 import org.gradle.workers.WorkerExecutor
 import org.jmailen.gradle.kotlinter.KotlinterExtension
 import org.jmailen.gradle.kotlinter.support.ExecutionContextRepository
+import org.jmailen.gradle.kotlinter.support.resolveRuleSets
 import org.jmailen.gradle.kotlinter.tasks.format.FormatExecutionContext
 import org.jmailen.gradle.kotlinter.tasks.format.FormatWorkerConfigurationAction
 import org.jmailen.gradle.kotlinter.tasks.format.FormatWorkerParameters
@@ -39,8 +40,9 @@ open class FormatTask @Inject constructor(
 
     @TaskAction
     fun run() {
+        val ruleSets = resolveRuleSets(experimentalRules)
         val executionContextRepository = ExecutionContextRepository.formatInstance
-        val executionContext = FormatExecutionContext(logger)
+        val executionContext = FormatExecutionContext(ruleSets, logger)
         val executionContextRepositoryId = executionContextRepository.register(executionContext)
 
         source

--- a/src/main/kotlin/org/jmailen/gradle/kotlinter/tasks/LintTask.kt
+++ b/src/main/kotlin/org/jmailen/gradle/kotlinter/tasks/LintTask.kt
@@ -14,8 +14,8 @@ import org.gradle.workers.WorkerExecutor
 import org.jmailen.gradle.kotlinter.KotlinterExtension
 import org.jmailen.gradle.kotlinter.support.HasErrorReporter
 import org.jmailen.gradle.kotlinter.support.ExecutionContextRepository
+import org.jmailen.gradle.kotlinter.support.defaultProviders
 import org.jmailen.gradle.kotlinter.support.reporterFor
-import org.jmailen.gradle.kotlinter.support.resolveRuleSets
 import org.jmailen.gradle.kotlinter.tasks.lint.LintExecutionContext
 import org.jmailen.gradle.kotlinter.tasks.lint.LintWorkerConfigurationAction
 import org.jmailen.gradle.kotlinter.tasks.lint.LintWorkerParameters
@@ -56,13 +56,12 @@ open class LintTask @Inject constructor(
 
     @TaskAction
     fun run() {
-        val ruleSets = resolveRuleSets(experimentalRules)
         val hasErrorReporter = HasErrorReporter()
         val reporters = reports.map { (reporter, report) ->
             reporterFor(reporter, report)
         } + hasErrorReporter
         val executionContextRepository = ExecutionContextRepository.lintInstance
-        val executionContextRepositoryId = executionContextRepository.register(LintExecutionContext(ruleSets, reporters, logger))
+        val executionContextRepositoryId = executionContextRepository.register(LintExecutionContext(defaultProviders(), reporters, logger))
 
         reporters.onEach { it.beforeAll() }
 

--- a/src/main/kotlin/org/jmailen/gradle/kotlinter/tasks/LintTask.kt
+++ b/src/main/kotlin/org/jmailen/gradle/kotlinter/tasks/LintTask.kt
@@ -15,6 +15,7 @@ import org.jmailen.gradle.kotlinter.KotlinterExtension
 import org.jmailen.gradle.kotlinter.support.HasErrorReporter
 import org.jmailen.gradle.kotlinter.support.ExecutionContextRepository
 import org.jmailen.gradle.kotlinter.support.reporterFor
+import org.jmailen.gradle.kotlinter.support.resolveRuleSets
 import org.jmailen.gradle.kotlinter.tasks.lint.LintExecutionContext
 import org.jmailen.gradle.kotlinter.tasks.lint.LintWorkerConfigurationAction
 import org.jmailen.gradle.kotlinter.tasks.lint.LintWorkerParameters
@@ -55,12 +56,13 @@ open class LintTask @Inject constructor(
 
     @TaskAction
     fun run() {
+        val ruleSets = resolveRuleSets(experimentalRules)
         val hasErrorReporter = HasErrorReporter()
         val reporters = reports.map { (reporter, report) ->
             reporterFor(reporter, report)
         } + hasErrorReporter
         val executionContextRepository = ExecutionContextRepository.lintInstance
-        val executionContextRepositoryId = executionContextRepository.register(LintExecutionContext(reporters, logger))
+        val executionContextRepositoryId = executionContextRepository.register(LintExecutionContext(ruleSets, reporters, logger))
 
         reporters.onEach { it.beforeAll() }
 

--- a/src/main/kotlin/org/jmailen/gradle/kotlinter/tasks/format/FormatExecutionContext.kt
+++ b/src/main/kotlin/org/jmailen/gradle/kotlinter/tasks/format/FormatExecutionContext.kt
@@ -1,6 +1,6 @@
 package org.jmailen.gradle.kotlinter.tasks.format
 
-import com.pinterest.ktlint.core.RuleSet
+import com.pinterest.ktlint.core.RuleSetProvider
 import org.gradle.api.logging.Logger
 import org.jmailen.gradle.kotlinter.support.ExecutionContext
 import java.util.Queue
@@ -10,7 +10,7 @@ import java.util.concurrent.ConcurrentLinkedQueue
  * Execution context for the formatting task.
  */
 data class FormatExecutionContext(
-    val ruleSets: List<RuleSet>,
+    val ruleSetProviders: Iterable<RuleSetProvider>,
     override val logger: Logger,
     val fixes: Queue<String> = ConcurrentLinkedQueue<String>()
 ) : ExecutionContext

--- a/src/main/kotlin/org/jmailen/gradle/kotlinter/tasks/format/FormatExecutionContext.kt
+++ b/src/main/kotlin/org/jmailen/gradle/kotlinter/tasks/format/FormatExecutionContext.kt
@@ -1,5 +1,6 @@
 package org.jmailen.gradle.kotlinter.tasks.format
 
+import com.pinterest.ktlint.core.RuleSet
 import org.gradle.api.logging.Logger
 import org.jmailen.gradle.kotlinter.support.ExecutionContext
 import java.util.Queue
@@ -9,6 +10,7 @@ import java.util.concurrent.ConcurrentLinkedQueue
  * Execution context for the formatting task.
  */
 data class FormatExecutionContext(
+    val ruleSets: List<RuleSet>,
     override val logger: Logger,
     val fixes: Queue<String> = ConcurrentLinkedQueue<String>()
 ) : ExecutionContext

--- a/src/main/kotlin/org/jmailen/gradle/kotlinter/tasks/format/FormatWorkerRunnable.kt
+++ b/src/main/kotlin/org/jmailen/gradle/kotlinter/tasks/format/FormatWorkerRunnable.kt
@@ -5,6 +5,7 @@ import com.pinterest.ktlint.core.RuleSet
 import org.gradle.api.logging.LogLevel
 import org.gradle.api.logging.Logger
 import org.jmailen.gradle.kotlinter.support.ExecutionContextRepository
+import org.jmailen.gradle.kotlinter.support.resolveRuleSets
 import org.jmailen.gradle.kotlinter.support.userData
 import java.io.File
 import javax.inject.Inject
@@ -40,7 +41,8 @@ class FormatWorkerRunnable @Inject constructor(
                         null
                     }
                 }?.let { formatFunc ->
-                    val formattedText = formatFunc.invoke(file, executionContext.ruleSets) { line, col, detail, corrected ->
+                    val ruleSets = resolveRuleSets(executionContext.ruleSetProviders, experimentalRules)
+                    val formattedText = formatFunc.invoke(file, ruleSets) { line, col, detail, corrected ->
                         val errorStr = "$relativePath:$line:$col: $detail"
                         val msg = when (corrected) {
                             true -> "Format fixed > $errorStr"

--- a/src/main/kotlin/org/jmailen/gradle/kotlinter/tasks/format/FormatWorkerRunnable.kt
+++ b/src/main/kotlin/org/jmailen/gradle/kotlinter/tasks/format/FormatWorkerRunnable.kt
@@ -5,7 +5,6 @@ import com.pinterest.ktlint.core.RuleSet
 import org.gradle.api.logging.LogLevel
 import org.gradle.api.logging.Logger
 import org.jmailen.gradle.kotlinter.support.ExecutionContextRepository
-import org.jmailen.gradle.kotlinter.support.resolveRuleSets
 import org.jmailen.gradle.kotlinter.support.userData
 import java.io.File
 import javax.inject.Inject
@@ -41,7 +40,7 @@ class FormatWorkerRunnable @Inject constructor(
                         null
                     }
                 }?.let { formatFunc ->
-                    val formattedText = formatFunc.invoke(file, resolveRuleSets(experimentalRules)) { line, col, detail, corrected ->
+                    val formattedText = formatFunc.invoke(file, executionContext.ruleSets) { line, col, detail, corrected ->
                         val errorStr = "$relativePath:$line:$col: $detail"
                         val msg = when (corrected) {
                             true -> "Format fixed > $errorStr"

--- a/src/main/kotlin/org/jmailen/gradle/kotlinter/tasks/lint/LintExecutionContext.kt
+++ b/src/main/kotlin/org/jmailen/gradle/kotlinter/tasks/lint/LintExecutionContext.kt
@@ -1,7 +1,7 @@
 package org.jmailen.gradle.kotlinter.tasks.lint
 
 import com.pinterest.ktlint.core.Reporter
-import com.pinterest.ktlint.core.RuleSet
+import com.pinterest.ktlint.core.RuleSetProvider
 import org.gradle.api.logging.Logger
 import org.jmailen.gradle.kotlinter.support.ExecutionContext
 
@@ -9,7 +9,7 @@ import org.jmailen.gradle.kotlinter.support.ExecutionContext
  * Execution context for the linting task.
  */
 data class LintExecutionContext(
-    val ruleSets: List<RuleSet>,
+    val ruleSetProviders: Iterable<RuleSetProvider>,
     val reporters: List<Reporter>,
     override val logger: Logger
 ) : ExecutionContext

--- a/src/main/kotlin/org/jmailen/gradle/kotlinter/tasks/lint/LintExecutionContext.kt
+++ b/src/main/kotlin/org/jmailen/gradle/kotlinter/tasks/lint/LintExecutionContext.kt
@@ -1,6 +1,7 @@
 package org.jmailen.gradle.kotlinter.tasks.lint
 
 import com.pinterest.ktlint.core.Reporter
+import com.pinterest.ktlint.core.RuleSet
 import org.gradle.api.logging.Logger
 import org.jmailen.gradle.kotlinter.support.ExecutionContext
 
@@ -8,6 +9,7 @@ import org.jmailen.gradle.kotlinter.support.ExecutionContext
  * Execution context for the linting task.
  */
 data class LintExecutionContext(
+    val ruleSets: List<RuleSet>,
     val reporters: List<Reporter>,
     override val logger: Logger
 ) : ExecutionContext

--- a/src/main/kotlin/org/jmailen/gradle/kotlinter/tasks/lint/LintWorkerRunnable.kt
+++ b/src/main/kotlin/org/jmailen/gradle/kotlinter/tasks/lint/LintWorkerRunnable.kt
@@ -6,6 +6,7 @@ import com.pinterest.ktlint.core.Reporter
 import com.pinterest.ktlint.core.RuleSet
 import org.gradle.api.logging.Logger
 import org.jmailen.gradle.kotlinter.support.ExecutionContextRepository
+import org.jmailen.gradle.kotlinter.support.resolveRuleSets
 import org.jmailen.gradle.kotlinter.support.userData
 import java.io.File
 import javax.inject.Inject
@@ -43,7 +44,8 @@ class LintWorkerRunnable @Inject constructor(
                     }
                 }
 
-                lintFunc?.invoke(file, executionContext.ruleSets) { error ->
+                val ruleSets = resolveRuleSets(executionContext.ruleSetProviders, experimentalRules)
+                lintFunc?.invoke(file, ruleSets) { error ->
                     reporters.onEach { it.onLintError(relativePath, error, false) }
 
                     val errorStr = "$relativePath:${error.line}:${error.col}: ${error.detail}"

--- a/src/main/kotlin/org/jmailen/gradle/kotlinter/tasks/lint/LintWorkerRunnable.kt
+++ b/src/main/kotlin/org/jmailen/gradle/kotlinter/tasks/lint/LintWorkerRunnable.kt
@@ -6,7 +6,6 @@ import com.pinterest.ktlint.core.Reporter
 import com.pinterest.ktlint.core.RuleSet
 import org.gradle.api.logging.Logger
 import org.jmailen.gradle.kotlinter.support.ExecutionContextRepository
-import org.jmailen.gradle.kotlinter.support.resolveRuleSets
 import org.jmailen.gradle.kotlinter.support.userData
 import java.io.File
 import javax.inject.Inject
@@ -44,7 +43,7 @@ class LintWorkerRunnable @Inject constructor(
                     }
                 }
 
-                lintFunc?.invoke(file, resolveRuleSets(experimentalRules)) { error ->
+                lintFunc?.invoke(file, executionContext.ruleSets) { error ->
                     reporters.onEach { it.onLintError(relativePath, error, false) }
 
                     val errorStr = "$relativePath:${error.line}:${error.col}: ${error.detail}"

--- a/src/test/kotlin/org/jmailen/gradle/kotlinter/support/ExecutionContextRepositoryTest.kt
+++ b/src/test/kotlin/org/jmailen/gradle/kotlinter/support/ExecutionContextRepositoryTest.kt
@@ -10,7 +10,7 @@ class ExecutionContextRepositoryTest {
     @Test
     fun getRegisteredContextWorks() {
         val repository = ExecutionContextRepository.lintInstance
-        val executionContext = LintExecutionContext(emptyList(), mock())
+        val executionContext = LintExecutionContext(mock(), emptyList(), mock())
         val id = repository.register(executionContext)
 
         val result = repository.get(id)
@@ -21,7 +21,7 @@ class ExecutionContextRepositoryTest {
     @Test(expected = NoSuchElementException::class)
     fun getUnregisteredContextFails() {
         val repository = ExecutionContextRepository.lintInstance
-        val executionContext = LintExecutionContext(emptyList(), mock())
+        val executionContext = LintExecutionContext(mock(), emptyList(), mock())
         val id = repository.register(executionContext)
         repository.unregister(id)
 

--- a/src/test/kotlin/org/jmailen/gradle/kotlinter/support/RuleSetsTest.kt
+++ b/src/test/kotlin/org/jmailen/gradle/kotlinter/support/RuleSetsTest.kt
@@ -13,14 +13,14 @@ class RuleSetsTest {
 
     @Test
     fun `resolveRuleSets loads from classpath providers`() {
-        val result = resolveRuleSets()
+        val result = resolveRuleSets(defaultProviders())
 
         assertEquals(listOf("standard"), result.map { it.id })
     }
 
     @Test
     fun `resolveRuleSets loads from classpath providers including experimental rules`() {
-        val result = resolveRuleSets(includeExperimentalRules = true)
+        val result = resolveRuleSets(defaultProviders(), true)
 
         assertEquals(listOf("standard", "experimental"), result.map { it.id })
     }
@@ -40,7 +40,7 @@ class RuleSetsTest {
 
     @Test
     fun `test compatibility`() {
-        KtLint.lint("""fun someFunc() = """"", resolveRuleSets(), {})
+        KtLint.lint("""fun someFunc() = """"", resolveRuleSets(defaultProviders())) {}
     }
 }
 

